### PR TITLE
added Layer.add_callback

### PR DESCRIPTION
**Pull request by beasteers**
This is in reference to #30809

### API Changes:
- added `Layer.add_callback(callback: k.callbacks.Callback) -> None`
- added `Layer.callbacks -> List[k.callbacks.Callback]` returns callbacks added to this layer and all children layers.

### Under-the-hood Changes
- `model.fit` concatenates the `callbacks` kw param with all children callbacks.

That's it 😊

### CODE REVIEW

**Comment by gbaned** [comment]
@rchao Can you please take a look on this PR? Thanks!

File: tensorflow/python/keras/engine/base_layer.py [hunk 1]
```diff
@@ -1223,6 +1227,19 @@ def add_metric(self, value, aggregation=None, name=None):
       # Insert layers into the Keras Graph Network.
       self._graph_network_add_metric(value, aggregation, name)
 
+  @property
+  def callbacks(self):
+    return self._callbacks + self._gather_children_attribute('callbacks')
```
Comment (rchao) [inline tensorflow/python/keras/engine/base_layer.py]: `_gather_children_attribute` has been recently removed. Can you edit the PR to make it applicable? Thanks.

**Comment by gbaned** [comment]
@beasteers Could you please address the reviewer comments and resolve conflicts? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by rthadur** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

File: tensorflow/python/keras/engine/base_layer.py (hunk 1 above)
Comment (beasteers) [reply tensorflow/python/keras/engine/base_layer.py:1232]: Or now that I'm looking at it more, `self._flatten_layers()` looks like another option - maybe that's what you were talking about when you said the other method was removed?

Here it's used for `metrics`:
https://github.com/tensorflow/tensorflow/blob/v2.4.1/tensorflow/python/keras/engine/base_layer.py#L1616-L1639
